### PR TITLE
CORTX-30161 : CORTX cluster upgrade automation

### PR DESCRIPTION
**Pull request by gauravchaudhari02**
# Problem Statement
- Automate upgrading CORTX cluster from old version to required/supported version using cortx-test framework

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
- [x] Jenkins pipelines used for testing
   cortx-upgrade - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/cortx-rgw-cluster-upgrade/3/
   cortx-upgrade with manually provided solution.yaml - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/cortx-rgw-cluster-upgrade-solution-input/9/
   nightly-cortx-upgrade - https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/cortx-rgw-cluster-nightly-upgrade/5/

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by gauravchaudhari02** [comment]
Successful email template for upgrade -
image
Failed email template for upgrade - 
image

File: jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        timeout(time: 240, unit: 'MINUTES')
+        timestamps()
+        buildDiscarder(logRotator(daysToKeepStr: '30', numToKeepStr: '30'))
+        ansiColor('xterm')
+    }
+
+    parameters {
+        string(name: 'CORTX_RE_BRANCH', defaultValue: 'main', description: 'Branch or GitHash for Cluster Setup scripts', trim: true)
+        string(name: 'CORTX_RE_REPO', defaultValue: 'https://github.com/Seagate/cortx-re', description: 'Repository for Cluster Setup scripts', trim: true)
+        string(name: 'CORTX_SERVER_IMAGE', defaultValue: 'ghcr.io/seagate/cortx-rgw:2.0.0-latest', description: 'CORTX-SERVER image', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy:18]: Please update the description to `CORTX-SERVER image to be updated`. Please update all image input parameters.

File: jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        string(name: 'CORTX_DATA_IMAGE', defaultValue: 'ghcr.io/seagate/cortx-data:2.0.0-latest', description: 'CORTX-DATA image', trim: true)
+        string(name: 'CORTX_CONTROL_IMAGE', defaultValue: 'ghcr.io/seagate/cortx-control:2.0.0-latest', description: 'CORTX-CONTROL image', trim: true)
+        choice(
+            name: 'POD_TYPE',
+            choices: ['all', 'data', 'control', 'ha', 'server'],
+            description: 'Pods need to be upgraded. option all will upgrade all CORTX services pods'
+        )
+        choice(
+            name: 'DEPLOYMENT_METHOD',
```
Comment (shailesh-vaidya) [inline jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy:27]: I guess we agreed to remove this parameter

File: jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                    env.changeset_log_url = changelog.absoluteUrl
+                    copyArtifacts filter: 'CHANGESET.txt', fingerprintArtifacts: true, flatten: true, optional: true, projectName: '/Release_Engineering/Cortx-Automation/changelog-generation', selector: lastCompleted(), target: ''
+                }
+            }
+        }
+
+        stage('QA SAnity') {
+            steps {
+                script {
+                    echo "Running QA Sanity"
```
Comment (shailesh-vaidya) [inline jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy:81]: Update statement to `QA Sanity job to be added`

File: solutions/kubernetes/cortx-deploy-functions.sh [hunk 4]
```diff
@@ -215,14 +215,12 @@ function update_solution_config(){
 }        
 
 function add_image_info() {
```
Comment (shailesh-vaidya) [inline solutions/kubernetes/cortx-deploy-functions.sh:217]: For this change please validate CORTX deployment both automated and manual

File: solutions/kubernetes/cortx-upgrade.sh [hunk 5]
```diff
@@ -0,0 +1,144 @@
+#!/bin/bash
+#
+# Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
```
Comment (shailesh-vaidya) [inline solutions/kubernetes/cortx-upgrade.sh:3]: s/2021/2022

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]

File: jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy (hunk 2 above)
Comment (gauravchaudhari02) [reply jenkins/automation/kubernetes/upgrade-cortx-cluster.groovy:27]: Yes. but we agreed that we will take it as improvement

File: jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy (hunk 3 above)
Comment (gauravchaudhari02) [reply jenkins/internal-ci/generic/nightly/nightly-upgrade.groovy]: Done

**Pull request merged**